### PR TITLE
Add Session Token to AWS IMDSv2 Metadata requests

### PR DESCRIPTION
**Pull request by sai-sunder-s**
AWS introduced IMDSv2 which brought in a requirement that a session token needs to be present as part of the metadata requests. This change uses a PUT request to fetch a session token and then uses the token in metadata requests to AWS.

Design Doc: go/aws-idmsv2 (Google Internal)

Note: This PR works on top of the changes made in this https://github.com/grpc/grpc/pull/24733

@markdroth

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/29296)

### CODE REVIEW

File: src/core/lib/security/credentials/external/aws_external_account_credentials.cc [hunk 1]
```diff
@@ -35,6 +35,10 @@ const char* kDefaultRegionEnvVar = "AWS_DEFAULT_REGION";
 const char* kAccessKeyIdEnvVar = "AWS_ACCESS_KEY_ID";
 const char* kSecretAccessKeyEnvVar = "AWS_SECRET_ACCESS_KEY";
 const char* kSessionTokenEnvVar = "AWS_SESSION_TOKEN";
+// const char* kImdsV2SessionTokenHeader = "x-aws-ec2-metadata-token";
+// const char* kImdsV2SessionTokenTTLHeader =
+// "x-aws-ec2-metadata-token-ttl-seconds"; const char* kImdsV2SessionTokenTTL =
+// "300";
```
Comment (renkelvin) [inline src/core/lib/security/credentials/external/aws_external_account_credentials.cc:41]: Please clean this up.

File: src/core/lib/security/credentials/external/aws_external_account_credentials.cc [hunk 2]
```diff
@@ -133,13 +143,83 @@ void AwsExternalAccountCredentials::RetrieveSubjectToken(
...
+      std::string(ctx_->response.body, ctx_->response.body_length);
+
   if (signer_ != nullptr) {
     BuildSubjectToken();
   } else {
     RetrieveRegion();
   }
 }
 
+void AwsExternalAccountCredentials::AddHeaders(grpc_http_request* request) {
```
Comment (renkelvin) [inline src/core/lib/security/credentials/external/aws_external_account_credentials.cc:212]: I think we should change this to a more specific name, `AddImdsHeader` for example.

File: src/core/lib/http/parser.cc [hunk 3]
```diff
@@ -361,6 +361,13 @@ void grpc_http_request_destroy(grpc_http_request* request) {
   gpr_free(request->path);
 }
 
+void grpc_http_header_destroy(grpc_http_header* header) {
+  if (header != nullptr) {
+    gpr_free(header->key);
+    gpr_free(header->value);
+  }
+}
+
```
Comment (renkelvin) [inline src/core/lib/http/parser.cc:370]: Where is this method used? The headers are already cleaned in the request/response destroy.

File: src/core/lib/security/credentials/external/aws_external_account_credentials.cc (hunk 2 above)
Comment (sai-sunder-s) [reply src/core/lib/security/credentials/external/aws_external_account_credentials.cc]: Renamed to AddMetadataRequestHeaders to make it consistent with other libraries

**Review by renkelvin — APPROVED** [review]

**Review by markdroth — CHANGES_REQUESTED** [review]
Looks good overall, just a few minor comments.  Once you address these, I'll trigger the tests.

Please let me know if you have any questions.  Thanks!

> Reviewed 6 of 10 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/29296#-:-N-UC0ca8Pap6xK9FCeR:bkowlb3)* status: all files reviewed, 6 unresolved discussions (waiting on @renkelvin and @sai-sunder-s)

___
*[src/core/lib/security/credentials/external/aws_external_account_credentials.cc, line 142 at r2](https://reviewable.io/reviews/grpc/grpc/29296#-N-U1urAEzWFfKZI1fLr:-N-U1urAEzWFfKZI1fLs:b-d4f1w6) ([raw file](https://github.com/grpc/grpc/blob/68646211c5e36c60a91d6004cd215b8fd86e7b29/src/core/lib/security/credentials/external/aws_external_account_credentials.cc#L142)):*
> ```cpp
>   ctx_ = ctx;
>   cb_ = cb;
>   if (imdsv2_session_token_url_.empty()) {
> ```

Suggest writing this as:

```
if (!imdsv2_session_token_url_.empty()) {
  RetrieveImdsV2SessionToken();
} else if (signer_ != nullptr) {
  BuildSubjectToken();
} else {
  RetrieveRegion();
}
```

That way, it's clearer what the precedence is of the three options.
___
*[src/core/lib/security/credentials/external/aws_external_account_credentials.cc, line 200 at r2](https://reviewable.io/reviews/grpc/grpc/29296#-N-U1zvm3jcqj2y8lH38:-N-U1zvn95Hie1OjzV27:bof0xp7) ([raw file](https://github.com/grpc/grpc/blob/68646211c5e36c60a91d6004cd215b8fd86e7b29/src/core/lib/security/credentials/external/aws_external_account_credentials.cc#L200)):*
> ```cpp
>   imdsv2_session_token_ =
>       std::string(ctx_->response.body, ctx_->response.body_length);
> ```

Please don't add blank lines within functions.
___
*[src/core/lib/security/credentials/external/aws_external_account_credentials.cc, line 211 at r2](https://reviewable.io/reviews/grpc/grpc/29296#-N-UAa2O4kGYbwbqxqyF:-N-UAa2P0TgHcOxfNbRi:b-xxrxr4) ([raw file](https://github.com/grpc/grpc/blob/68646211c5e36c60a91d6004cd215b8fd86e7b29/src/core/lib/security/credentials/external/aws_external_account_credentials.cc#L211)):*
> ```cpp
>     grpc_http_request* request) {
>   if (!imdsv2_session_token_.empty()) {
>     grpc_http_header* headers =
> ```

Let's make sure we're not accidentally overwriting any pre-existing headers here:

```
GPR_ASSERT(request->hdr_count == 0);
GPR_ASSERT(request->hdrs == nullptr);
```

**Review by sai-sunder-s — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/29296)* status: 9 of 10 files reviewed, 6 unresolved discussions (waiting on @markdroth, @renkelvin, and @sai-sunder-s)

___
*[src/core/lib/security/credentials/external/aws_external_account_credentials.cc, line 142 at r2](https://reviewable.io/reviews/grpc/grpc/29296#-N-U1urAEzWFfKZI1fLr:-N-_6Vce0UOCslW9uXkR:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/68646211c5e36c60a91d6004cd215b8fd86e7b29/src/core/lib/security/credentials/external/aws_external_account_credentials.cc#L142)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Suggest writing this as:

```
if (!imdsv2_session_token_url_.empty()) {
  RetrieveImdsV2SessionToken();
} else if (signer_ != nullptr) {
  BuildSubjectToken();
} else {
  RetrieveRegion();
}
```

That way, it's clearer what the precedence is of the three options.
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/external/aws_external_account_credentials.cc, line 200 at r2](https://reviewable.io/reviews/grpc/grpc/29296#-N-U1zvm3jcqj2y8lH38:-N-_6TS04hxmFRf9YSnM:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/68646211c5e36c60a91d6004cd215b8fd86e7b29/src/core/lib/security/credentials/external/aws_external_account_credentials.cc#L200)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Please don't add blank lines within functions.
</blockquote></details>

Done.
___
*[src/core/lib/security/credentials/external/aws_external_account_credentials.cc, line 211 at r2](https://reviewable.io/reviews/grpc/grpc/29296#-N-UAa2O4kGYbwbqxqyF:-N-_6PprDnJlFQJIxrN2:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/68646211c5e36c60a91d6004cd215b8fd86e7b29/src/core/lib/security/credentials/external/aws_external_account_credentials.cc#L211)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Let's make sure we're not accidentally overwriting any pre-existing headers here:

```
GPR_ASSERT(request->hdr_count == 0);
GPR_ASSERT(request->hdrs == nullptr);
```
</blockquote></details>

Done.

**Review by markdroth — APPROVED** [review]
Looks great!  I've triggered the tests, let's see what they show.

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc/29296#-:-N-_7wEy0WS4kmLQ5Vw3:bf0asow)* status: all files reviewed, 3 unresolved discussions (waiting on @renkelvin and @sai-sunder-s)

**Comment by renkelvin** [comment]
Thanks Sai and thanks Mark!

**Pull request merged**